webapp/lib/controller.dart

```diff
@@ -169,59 +131,102 @@ void loadDashboardView() {
-  // view.contentView.batchRepliesConfigurationView.tagResponses.renderResponses(selectedTag, responseLanguages, tagResponses);
-
-  // view.contentView.escalatesConfigurationView.tagList.renderTagList(tagList);
-  // view.contentView.escalatesConfigurationView.tagResponses.renderResponses(selectedTag, responseLanguages, tagResponses);
+void hasAllTagsChanged (String tag, model.TagStyle tagStyle) {
+  _updateTag(model.changeCommsPackage.hasAllTags, tag, tagStyle);
+  loadBatchRepliesConfigurationView();
 }
 
-void addNewConfigurationTag(String tagToAdd, List<String> availableLanguages, Set<String> additionalTags, Map<String, List<List<String>>> tagData) {
-  tagData[tagToAdd] = [configurationResponseLanguages.map((e) => '').toList()];
-  additionalTags.remove(tagToAdd);
-  populateConfigurationView(tagToAdd, getTagList(tagToAdd, tagData), availableLanguages, tagData[tagToAdd]);
+void containsLastInTurnTagsChanged (String tag, model.TagStyle tagStyle) {
+   _updateTag(model.changeCommsPackage.containsLastInTurnTags, tag, tagStyle);
+  loadBatchRepliesConfigurationView();
 }
 
-void updateEditedConfigurationTagResponse(String parentTag, int textIndex, int languageIndex, String text, List<List<String>> tagResponses) {
-  tagResponses[textIndex][languageIndex] = text;
+void hasNoneTagsChanged (String tag, model.TagStyle tagStyle) {
+   _updateTag(model.changeCommsPackage.hasNoneTags, tag, tagStyle);
+  loadBatchRepliesConfigurationView();
 }
 
-void addConfigurationResponseEntries(String parentTag, int languageIndex, String text, List<String> responseLanguages, Map<String, List<List<String>>> tagData) {
-  if (languageIndex != null && text != null) {
-    var pos = tagData[parentTag].indexWhere((x)=> x.contains(''));
-    if (pos > -1) {
-      tagData[parentTag][pos][languageIndex] = text;
+void addsTagsChanged (String updatedTag, String originalTag, model.TagStyle tagStyle) {
+  if (model.changeCommsPackage.addsTags.containsKey(originalTag)) {
+    if (updatedTag != originalTag) {
+      var addsTagsKeys = model.changeCommsPackage.addsTags.keys.toList();
+      var addsTagsTagValues= model.changeCommsPackage.addsTags.values.toList();
+      var originalIndex = addsTagsKeys.indexOf(originalTag);
+      addsTagsKeys.removeAt(originalIndex);
+      addsTagsKeys.insert(originalIndex, updatedTag);
+      Map<String, model.TagStyle> updatedAddsTags = {};
+      for (int i = 0; i < addsTagsKeys.length; i++) {
+        updatedAddsTags[addsTagsKeys[i]] = addsTagsTagValues[i];
+      }
+      model.changeCommsPackage.addsTags = updatedAddsTags;
     } else {
-      tagData[parentTag].add(configurationResponseLanguages.map((e) => '').toList());
-      tagData[parentTag].last[languageIndex]= text;
+      model.changeCommsPackage.addsTags.remove(originalTag);
     }
   } else {
-    tagData[parentTag].add(['', '']);
+    model.changeCommsPackage.addsTags.addAll({updatedTag: tagStyle});
   }
```

Comment:
In general, multi-level if-else code is going to be tricky to maintain, and error prone. There's also some overlap here with the `_updateTag` method. We know in the view when the user adds/removes/changes a tag since the event we need to listen to is different, so I suggest creating an enum with the three possible types of changes and passing that from the view to all these "...TagsChanged" callbacks.

Reply:
I've come up with a better structure with the use of an enum